oauthproxy.go

```diff
@@ -633,7 +633,30 @@ func getRemoteAddr(req *http.Request) (s string) {
 	return
 }
 
+// See https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/http-caching?hl=en
+var noCacheHeaders = map[string]string{
+	"Expires":         time.Unix(0, 0).Format(time.RFC1123),
+	"Cache-Control":   "no-cache, no-store, must-revalidate, max-age=0",
+	"X-Accel-Expires": "0", // https://www.nginx.com/resources/wiki/start/topics/examples/x-accel/
+}
+
+// prepareNoCache prepares headers for preventing browser caching.
+func prepareNoCache(w http.ResponseWriter) {
+	// Set NoCache headers
+	for k, v := range noCacheHeaders {
+		w.Header().Set(k, v)
+	}
+}
+
+func hasProxyPrefix(path, prefix string) bool {
+	return strings.HasPrefix(path, prefix)
+}
+
 func (p *OAuthProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	if hasProxyPrefix(req.URL.Path, p.ProxyPrefix) {
+		prepareNoCache(rw)
```

Comment:
In "proxy" mode (in contrast to "auth_request" mode), oauth2-proxy serves the sign-in page directly when you try to access some other url and are not authenticated, or it serves the redirect to the provider to start the oauth process. These responses are to the application URL and do not have the proxy prefix, but we also want to prevent the browser caching these. (Even if oauth2-proxy were to serve a redirect to its own path first, we would still want to prevent that redirect from being cached.)

So, a possible fix would be to add these headers at the start of the `OAuthProxy.SignInPage()` and `OAuthProxy.OAuthStart()` methods. There might be more thorough or elegant strategies, but I'm not sure ...

Reply:
Normally, we should use http middleware to do those process, but the structure of oauth2-proxy's http handler is a bit complicated, so it's difficult.
For now, I added prepareNoCache processing to SignInPage and OAuthStart.

Reply:
looks good to me